crates/matrix-sdk/src/widget_api/permissions.rs

```diff
@@ -0,0 +1,24 @@
+//! Types and traits related to the permissions that a widget can request from a client.
+
+use async_trait::async_trait;
+
+/// A trait that must be implemented by an entity that performs the validation / authorization
+/// of the widget's permissions. Typically, it's a native client that "hosts" the widget.
+#[async_trait]
+pub trait PermissionsProvider: Send + Sync + 'static {
+    /// Receives a request for given permissions and returns the actual permissions that
+    /// the user (client) granted to a given widget.
+    async fn acquire_permissions(&self, permissions: Permissions) -> Permissions;
+}
+
+/// Permissions that a widget can request from a client.
+#[derive(Debug)]
+pub struct Permissions {
+    /// Types of the messages that a widget wants to be able to fetch.
+    pub read: Vec<MessageType>,
+    /// Types of the messages that a widget wants to be able to send.
+    pub send: Vec<MessageType>,
+}
+
+/// An alias for an actual type of the messages that is going to be used with a permission request.
+pub type MessageType = String;
```

Comment:
This is for a Matrix event type, and widgets might interact with both message events and state events, right? If yes, this type alias doesn't need to exist, we can use `TimelineEventType` from `ruma::events`.

Reply:
Right (if I remember correctly). I actually wanted to also use stronger types here and had a discussion with @toger5, but as far as I can remember, @toger5's suggestion was that it can be any event type in theory and that there was no simple way to map them (no enum with a function `f: EventType -> String` and `f: String -> EventType` transformation). @toger5, could you chime in? 🙂

Reply:
There are other "event" types that aren't part of `TimelineEventType`, but I'm 99.5% certain that those are out of scope for the widget API. They are ephemeral or mutable things like ephemeral room events (typing, receipts), account data and presence. Often when people talk about events, those are implicitly excluded (see also https://github.com/matrix-org/matrix-spec/issues/897).

Reply:
Widgets are allowed to introduce new unspecced message types. Just arbitrary strings for their own usecases. Does TimelineEevntType have a "other" enum case. If so this would very much be a better solution. If not it would be a shame if a widget is working with its own timeline event types but the widget driver is giving a parsing error.

Reply:
> Does TimelineEevntType have a "other" enum case.

Yes,any string can be converted `.into()` that type. The variant is hidden / opaque, but to check for a custom type, one can also convert the other way using `.to_string()`.

Reply:
Nice! Updated.